feat(trace): Spec Neotrace - automatic test tracing with smart state tracking

IvanAnishchuk opened the pull request:
<!-- Description
Provide at least one paragraph that clearly and succinctly explains:
* What this PR does (but not how it does it)
* Why this PR is necessary, including context
-->

This PR provides a major part of a new testing framework - spec wrapper with automatic test execution tracing and smart beacon state tracking. It's based on #4603 and lessons learned in #4724 were taken into account.

This is fully backwards-compatible thing, new test/infra module was added, no preexisting code changed.

Key Changes:

Smart State Tracking: The RecordingSpec now automatically detects state context switches by tracking the root hash of the state argument. It prepends a load_state operation only when the state has actually changed (including out-of-band mutations in tests), removing the need for manual yields and the like.

Pydantic Serialization: Leverages Pydantic field_validator and model_dump to automatically handle type coercion (e.g., converting bytes to hex strings, int subclasses to primitives) and sanitation.

Lean Proxy: Reduced RecordingSpec (based on wrapt) to a thin wrapper that strictly handles function interception and flow control.

<!-- Checklist
Ensure the following tasks have been done prior to submitting the PR:
* Update documentation (if applicable)
* Add tests for new functionality (if applicable)
* Run `make lint` to check formatting
* Run `make test` to check tests
-->


Testing: Added unit tests for the new tools and example spec tests based on the new infra. All tests are passing, lint was done.


<!-- Relations
Link any related PRs or issues:
* Use "Fixes #123" to auto-close issues
* Use "Related to #456" for related work
-->

Fixes #4603
Related to #4724

IvanAnishchuk committed:
feat(trace): spec tracing stage 1 prototype

pydantic models for the spec trace
core spec tracing logic
use wrapt to wrap the spec and intercept the calls
tracing decorator
some basic unit tests for the trace recorder
some converted test examples
use 0x prefix for hex bytes in trace
a README with a short explanation how tracing works

leolara commented on tests/infra/trace/models.py:
```diff
@@ -0,0 +1,204 @@
+
+        return context_name
+
+    def dump_to_dir(self, output_dir: str, config: dict[str, Any] = None) -> None:
```
Is the normal pydantic way of saving to have this as a method of the model, or having externally in another function? We should do it the pydantic way

IvanAnishchuk replied:
Let me check... I'm more used to seeing it as a method but let's do idiomatic :+1:

leolara commented on tests/infra/trace/decorator.py:
```diff
@@ -0,0 +1,143 @@
+
+    def decorator(fn: Callable):
+        @functools.wraps(fn)
+        def wrapper(*args, **kwargs):
```
Not compatible with the dumper. I think we need to return the pydantic model. Then make sure that parts that expect the yields up the calling stack also can let this type pass through. Then in this function: https://github.com/ethereum/consensus-specs/blob/master/tests/core/pyspec/eth2spec/gen_helpers/gen_base/gen_runner.py#L87 we handle differently if it is returned a iterator/generator or the pydantic mode. Then there if there is a pydantic model we dump it. In this function `test_case` contains all the meta info about the test we are running, so that way you can calculate the folder.

IvanAnishchuk committed:
feat(trace): spec tracing stage 2 partial rewrite and cleanup

add "method" to StepModel for spec_call op
remove unneeded things
address some more requirements, format, etc.
new approach - decorator just generates data for dumper
add the auto-assert of state in the end of test trace
adjust assert/load tracing logic according to the issue
rename record_spec_trace -> spec_trace
test fixes
more simplicity
some cleanup

IvanAnishchuk committed:
feat(trace): tracing stage 3 minor fixes

Draft a naive approach to trace data dumping.
Small fixes and lots of polish.
Test improvements.

IvanAnishchuk commented on traces/.gitignore:
```diff
@@ -0,0 +1,2 @@
+*
+!.gitignore
```
Probably can remove this

IvanAnishchuk commented on tests/infra/trace/traced_spec.py:
```diff
@@ -0,0 +1,195 @@
+
+    def _capture_post_state(
+        self,
+        state: View,
```
View | None

IvanAnishchuk commented:
@leolara if you could just take another glance? :)

I'm not sure about the best way to integrate object-returning tests into the same harness as generator tests. I did something and it works with reftests and has minimal impact on existing code but it doesn't look quite right, yet I don't want to rewrite half the test framework for supporting this either, at least not without some guidance.

IvanAnishchuk committed:
feat(trace): tracing stage 4 minor style improvements

Add shared base class for Assert/Load state ops.
Stricter and more concise typing.
Simplify serialization.

IvanAnishchuk committed:
feat(trace): tracing stage 5 some more style improvements

Typing annotations, comments, edge cases.
Fix extra deps, fix styiling in the README.
